src/psyclone/tests/psyir/transformations/arrayrange2loop_trans_test.py

```diff
@@ -40,445 +40,355 @@
-    lhs = create_array_y_2d_slice(symbol_table)
-    rhs = create_array_z(symbol_table)
-    assignment = Assignment.create(lhs, rhs)
-    routine = KernelSchedule.create("work", symbol_table,
-                                    [assignment])
-    trans.apply(assignment)
-    trans.apply(assignment)
-    expected = (
-        "  do idx = LBOUND(y2, dim=2), UBOUND(y2, dim=2), 1\n"
-        "    do idx_1 = LBOUND(y2, dim=1), UBOUND(y2, dim=1), 1\n"
-        "      y2(idx_1,idx) = z(idx_1,n,idx)\n"
-        "    enddo\n"
-        "  enddo\n")
-    writer = FortranWriter()
-    result = writer(routine)
-    assert expected in result
-    assert Compile(tmpdir).string_compiles(result)
-
-
-def test_transform_apply_insert(tmpdir):
-    '''Check that the PSyIR is transformed as expected when there are
-    multiple statements in the PSyIR. The resultant Fortran code is used to
-    confirm the transformation has worked correctly.
+          ! Test 2: We know the bounds and they are not equal
+          x2(:,:) = y2(:,:)
 
-    '''
-    trans = ArrayRange2LoopTrans()
+          ! Test 3: We know the bounds and we have implicit bounds accesses
+          x2(2:4,4:6) = y2(15:17,28:30) + 1.0
 
-    symbol_table = SymbolTable()
-    symbol = DataSymbol("n", INTEGER_TYPE)
-    symbol_table.add(symbol)
-    # Create the first assignment. In Fortran notation: x(:) = y(n,:)
-    lhs = create_array_x(symbol_table)
-    rhs = create_array_y(symbol_table)
-    assignment1 = Assignment.create(lhs, rhs)
-    # Create the second assignment. In Fortran notation: y2(:,:) = z(:,n,:)
-    lhs = create_array_y_2d_slice(symbol_table)
-    rhs = create_array_z(symbol_table)
-    assignment2 = Assignment.create(lhs, rhs)
-    routine = KernelSchedule.create("work", symbol_table,
-                                    [assignment1, assignment2])
-    trans.apply(assignment1)
-    trans.apply(assignment2)
-    writer = FortranWriter()
-    expected = (
-        "  do idx = LBOUND(x, dim=1), UBOUND(x, dim=1), 1\n"
-        "    x(idx) = y(n,idx)\n"
-        "  enddo\n"
-        "  do idx_1 = LBOUND(y2, dim=2), UBOUND(y2, dim=2), 1\n"
-        "    y2(:,idx_1) = z(:,n,idx_1)\n"
-        "  enddo\n")
-    result = writer(routine)
-    assert expected in result
-    assert Compile(tmpdir).string_compiles(result)
+          ! Test 4: SoA and SoAoS
+          x(:) = struct%values(:) + struct%array(:)%value
+        end program test
+    ''')
+    trans = ArrayRange2LoopTrans()
+    for assignment in psyir.walk(Assignment):
+        trans.apply(assignment)
+
+    # The bounds are not known, so L/UBOUND expressions are used
+    output_test1 = fortran_writer(psyir.walk(Assignment)[0])
+    assert ("x1(idx_1,idx) = y1(idx_1 + (LBOUND(y1, dim=1) - LBOUND(x1, "
+            "dim=1)),idx + (LBOUND(y1, dim=2) - LBOUND(x1, dim=2)))"
+            in output_test1)
+
+    # When we know the bounds we can see they are different, we also
+    # need the offsets (and can also use L/UBOUND)
```

Comment:
Is the theory here that using LBOUND/UBOUND will make no difference as far as the compiler is concerned so it's not worth our while putting in the explicit values which we know from the declaration?

Reply:
yes, using the explicit values is not straight-forward here since we just know that we need offsets because the "same_range" could not be guaranteed. But we would need all the logic to check if each bound is known or not, if they are from unknown/unresolved type, if the shape is attribute/deferred, ... Unless we know this is an optimisation (which doesn't seem to be), the L/UBOUND works and are much simpler.